Tests/TensorFlowTests/EpochsTests.swift

```diff
@@ -0,0 +1,383 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import XCTest
+
+@testable import TensorFlow
+
+var rng = ARC4RandomNumberGenerator(seed: [42])
+
+final class EpochsTests: XCTestCase {
+ 
+  // An element that keeps track of when it was first accessed.
+  class AccessTracker {
+      var accessed: Bool = false
+  }
+  
+  // A struct keeping track of when its elements have been first accessed. We 
+  // use it in the tests to check whether methods that are not supposed to break
+  // the laziness work as intended.
+  struct ReadTracker<Base: RandomAccessCollection> : RandomAccessCollection {
+    let base: Base
+    let accessed_: [AccessTracker]
+      
+    public typealias Element = Base.Element
+    /// A type whose instances represent positions in `self`.
+    public typealias Index = Base.Index
+    /// The position of the first element.
+    public var startIndex: Index { base.startIndex }
+    /// The position one past the last element.
+    public var endIndex: Index { base.endIndex }
+    /// Returns the position after `i`.
+    public func index(after i: Index) -> Index { base.index(after: i) }
+    /// Returns the position after `i`.
+    public func index(before i: Index) -> Index { base.index(before: i) }
+
+    init(_ base: Base) { 
+      self.base = base
+      accessed_ = (0..<base.count).map { _ in AccessTracker() } 
+    }  
+    
+    subscript(i: Base.Index) -> Base.Element {
+      accessed_[base.distance(from: base.startIndex, to: i)].accessed = true
+      return base[i]
+    }
+    
+    var accessed: LazyMapCollection<[AccessTracker], Bool> {
+      accessed_.lazy.map(\.accessed)
+    }
+  }
+    
+  func testBaseUse() {
+    let batchSize = 64
+    let dataset = (0..<512).map { (_) -> Tensor<Float> in 
+      Tensor<Float>(randomNormal: [224, 224, 3]) 
+    } 
+    let batches = dataset.inBatches(of: batchSize).lazy.map(\.collated)
+
+    XCTAssertEqual(batches.count, dataset.count / batchSize,
+                   "Incorrect number of batches.")
+    for batch in batches {
+      XCTAssertEqual(batch.shape, TensorShape([64, 224, 224, 3]),
+        "Wrong shape for batch: \(batch.shape), should be [64, 224, 224, 3]")
+    }
+  }
+    
+  func testInBatchesIsLazy() {
+    let batchSize = 64
+    let items = Array(0..<512)
+    let dataset = ReadTracker(items)
+    let batches = dataset.inBatches(of: batchSize)
+      
+    // `inBatches` is lazy so no elements were accessed.
+    XCTAssert(dataset.accessed.allSatisfy(){ !$0 },
```

Comment:
Yeah, we should also have `.allTrue` and `.allFalse` in StdlibExtensions.swift to make things like this more expressive.  No need to do that now though.
```suggestion
    XCTAssert(dataset.accessed.allSatisfy { !$0 },
```

If you run swift-format on this it's going to insert a newline after the open paren.  The rule is, if an opening delimiter such as `(` can't be closed on the same line, the opening delimiter should be the last thing on its line.  There's one exception: an unmatched-on-the-same-line `{` can be followed by a closure parameter list.